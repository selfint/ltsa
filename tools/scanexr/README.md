# scanexr

Find call path from a known vulnerable value,
to a known controllable value.

A 'value' can be a method, an argument to a method, or anything else.
