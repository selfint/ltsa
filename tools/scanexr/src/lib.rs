use std::{fmt::Debug, path::Path};

use anyhow::{Context, Result};
use async_recursion::async_recursion;
use async_trait::async_trait;
use lsp_client::client::Client;
use step::Step;
use tree_sitter::{Language, Query};

use crate::utils::get_query_steps;

pub mod step;
pub mod tracers;
pub mod utils;

pub type Stacktrace<C> = Vec<Step<C>>;

#[async_trait]
pub trait Tracer: Send + Sync {
    type StepContext: Debug + Default + Clone + Send + Sync;

    fn get_language(&self) -> Language;

    /// given a step, get all possible stack traces leading to it
    async fn get_stacktraces(
        &self,
        lsp_client: &Client,
        step: &Step<Self::StepContext>,
        stop_at: &[Step<Self::StepContext>],
    ) -> Result<Vec<Stacktrace<Self::StepContext>>>;
}

pub async fn get_all_stacktraces<T: Tracer>(
    tracer: &T,
    lsp_client: &Client,
    root_dir: &Path,
    pub_queries: &[(Query, u32)],
    hacky_query: &(Query, u32),
) -> Result<Vec<Stacktrace<T::StepContext>>> {
    let mut pub_steps = vec![];
    for pub_query in pub_queries {
        let steps = get_query_steps(root_dir, tracer.get_language(), pub_query)
            .context("getting pub query steps")?;
        for step in steps {
            if !pub_steps.contains(&step) {
                pub_steps.push(step);
            }
        }
    }

    let hacky_steps = get_query_steps(root_dir, tracer.get_language(), hacky_query)
        .context("getting hacky steps")?;

    let mut all_stacktraces = vec![];
    for hacky_step in &hacky_steps {
        let stacktraces = _get_all_stacktraces(tracer, lsp_client, hacky_step, &pub_steps)
            .await
            .context("completing stacktraces")?;

        all_stacktraces.extend(stacktraces);
    }

    Ok(all_stacktraces)
}

#[async_recursion]
async fn _get_all_stacktraces<T: Tracer>(
    tracer: &T,
    lsp_client: &Client,
    step: &Step<T::StepContext>,
    stop_at: &[Step<T::StepContext>],
) -> Result<Vec<Stacktrace<T::StepContext>>> {
    // get stacktraces leading to step
    let stacktraces = tracer.get_stacktraces(lsp_client, step, stop_at).await?;

    // complete stacktraces leading to step
    let mut completed_stacktraces = vec![];
    for stacktrace in stacktraces {
        let next_stacktraces = _get_all_stacktraces(tracer, lsp_client, step, stop_at).await?;
        for next_stacktrace in next_stacktraces {
            let mut completed_stacktrace = stacktrace.clone();
            completed_stacktrace.extend(next_stacktrace);

            completed_stacktraces.push(completed_stacktrace);
        }
    }

    Ok(completed_stacktraces)
}
