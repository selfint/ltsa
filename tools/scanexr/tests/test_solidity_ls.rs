use scanexr::{
    language_provider::LspProvider, languages::solidity::SolidityLs, test_utils::display_locations,
    utils::visit_dirs,
};

#[tokio::test]
async fn test_find_definitions() {
    macro_rules! test_definitions {
        ($input:literal) => {
            let (root_dir, location, _, _) = scanexr::test_utils::setup_test_dir($input);

            let mut project_files = vec![];
            visit_dirs(root_dir.path(), &mut |f| project_files.push(f.path()))
                .expect("failed to get project files");

            let lsp = SolidityLs::new(root_dir.path(), project_files)
                .await
                .expect("failed to start solidity ls");

            let definitions = lsp
                .find_definitions(&location)
                .await
                .expect("failed to find definition");

            let definitions = definitions.into_iter().map(|d| (d, ())).collect();
            let snapshot = format!(
                r#"
### input ###
{}
### output ###
{}"#,
                $input,
                display_locations::<()>(definitions),
            );

            insta::assert_snapshot!(snapshot);
        };
    }

    test_definitions!(
        r#"
contract.sol
#@#
pragma solidity ^0.8.19;

contract Contract {
    function withdraw() public {
        address target = msg.sender;

        (bool sent, ) = target.call{value: 1}("");
                     // ^^^^^^ start
    }
}
        "#
    );

    test_definitions!(
        r#"
contract.sol
#@#
pragma solidity ^0.8.19;

contract Contract {
    function foo() public {}

    function withdraw() public {
        address target = msg.sender;
        
        foo();
    //  ^^^ start

        (bool sent, ) = target.call{value: 1}("");
    }
}
        "#
    );
}

#[tokio::test]
async fn test_find_references() {
    macro_rules! test_references {
        ($input:literal) => {
            let (root_dir, location, _, _) = scanexr::test_utils::setup_test_dir($input);

            let mut project_files = vec![];
            visit_dirs(root_dir.path(), &mut |f| project_files.push(f.path()))
                .expect("failed to get project files");

            let lsp = SolidityLs::new(root_dir.path(), project_files)
                .await
                .expect("failed to start solidity ls");

            let references = lsp
                .find_references(&location)
                .await
                .expect("failed to find references");

            let references = references.into_iter().map(|d| (d, ())).collect();
            let snapshot = format!(
                r#"
### input ###
{}
### output ###
{}"#,
                $input,
                display_locations::<()>(references),
            );

            insta::assert_snapshot!(snapshot);
        };
    }

    test_references!(
        r#"
contract.sol
#@#
pragma solidity ^0.8.19;

contract Contract {
    function foo() public {}
         //  ^^^ start

    function withdraw() public {
        address target = msg.sender;
        
        foo();
        foo();

        (bool sent, ) = target.call{value: 1}("");
    }
}
        "#
    );
}
