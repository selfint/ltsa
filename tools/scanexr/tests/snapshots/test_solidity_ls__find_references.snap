---
source: tools/scanexr/tests/test_solidity_ls.rs
expression: snapshot
---

### input ###

contract.sol
#@#
pragma solidity ^0.8.19;

contract Contract {
    function foo() public {}
         //  ^^^ start

    function withdraw() public {
        address target = msg.sender;
        
        foo();
        foo();

        (bool sent, ) = target.call{value: 1}("");
    }
}
        
### output ###
contract.sol
#@#

pragma solidity ^0.8.19;

contract Contract {
    function foo() public {}
         //  ^^^ start

    function withdraw() public {
        address target = msg.sender;
        
        foo();
        ^^^ Meta: ()
        foo();

        (bool sent, ) = target.call{value: 1}("");
    }
}
        
---
contract.sol
#@#

pragma solidity ^0.8.19;

contract Contract {
    function foo() public {}
         //  ^^^ start

    function withdraw() public {
        address target = msg.sender;
        
        foo();
        foo();
        ^^^ Meta: ()

        (bool sent, ) = target.call{value: 1}("");
    }
}
        
